### PR TITLE
[4.0] Null date corrections for frontent edit icon

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for Issue #26783 , see also [https://github.com/joomla/joomla-cms/pull/26751#issuecomment-544600907](https://github.com/joomla/joomla-cms/pull/26751#issuecomment-544600907) .

### Summary of Changes

Changes the code to produce the edit icon for frontend editing so it handles real null values for published up and down times.

This fixes the issue that on current 4.0-dev, they eye icon is shown for published articles instead of the edit icon, like it would be with unpublished or not published yet articles.

### Testing Instructions

Use an installation of current 4.0-dev or latest nightly build of today.

1. Create an article with state = published and show it on frontend.
2. On frontend, login as used who has permission for frontend editing.
3. Click the edit button (gear icon at the right top corner of article).
4. In the pulldown which appears below the gear icon, check the icon shown for the "Edit" menu item.

Result: See section "Actual result" below.

5. Apply patch.
6. Reload the page on frontend.
7. Check again the icon in the frontend editing pulldown menu.

Result: See section "Expected result" below.

8. Play around with publish up and down times so you have following situations, and check the icon again:

- Published down in past = article expired
- Published up in future = article pending (i.e. not published yet)
- Published up in past, published down in future = article published
- If you want play with other combinations with empty or not empty published up or down fields.

### Expected result

For published articles the edit icon is shown and for unpublished, pending or expired articles the eye icon.

snap-14

### Actual result

For published articles the same eye icon is shown as for unpublished, expired or pending articles.

snap-13

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by richard67** [comment]
@wilsonge Do we need to check old nulldates here for 3rd party compatibility? Is not a library, but can the layout used by 3rd party? As far as I can it is only made for articles, and those are created by core, so we don't need to check old nulldates, right? Or shall we better stay safe and keep the old nulldates like it is now in this PR?

**Comment by Stuartemk** [comment]
In pull 26751 @ infograf768 it says: Null Date is now defined by null and no more by Factory :: getDbo () -> getNullDate ()

Now in this pull it is used again ... Am I missing something or is there something I don't understand?

Or the problem is that there is no established parameter of code ?, as it still happens with the array limes () and that they should be replaced by [] but still use them in the development of Joomla 4.x ... as as many deprecated that are still written.

Thank you.

**Comment by richard67** [comment]
@Stuartemk We want to keep the old nulldates in addition to real nulls only where it might be necessary for 3rd party extensions. Here I am not sure if to keep it or not, that's why I've pinged wilsonge before for his opinion.

**Comment by infograf768** [comment]
@richard67 
> Or shall we better stay safe and keep the old nulldates like it is now in this PR?
 
That layout can be used by a 3rd party. Let's keep the old nulldates.

**Comment by infograf768** [comment]
I get with this patch depending on the state of the article:

Screen Shot 2019-10-22 at 07 23 11

and

Screen Shot 2019-10-22 at 07 22 30

Is that OK? (Except evidently for the display of the tips)

**Comment by richard67** [comment]
@infograf768 Yes, that’s how it shall be (except the bad tips of course).

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 28202f9d82a7deb4847607a650e15a3e4ba49091<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26756">issues.joomla.org/tracker/joomla-cms/26756</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 28202f9d82a7deb4847607a650e15a3e4ba49091<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26756">issues.joomla.org/tracker/joomla-cms/26756</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26756">issues.joomla.org/tracker/joomla-cms/26756</a>.</sub>

**Comment by SharkyKZ** [comment]
>Or shall we better stay safe and keep the old nulldates like it is now in this PR?

No. This layout is specific to com_content.

**Comment by richard67** [comment]
George confirmed in Glip that I shall keep the old nulldates, too, because this layout might be used by 3rd partries, too, and I think if not then it doesn't hurt. Naming of variables and so on look to me as if it was made for articles only, but this might be for historic reasons because frontend editing at the very first beginning was limited to articles only, if I remeber right. So all ok here.

File: layouts/joomla/content/icons/edit.php [hunk 1]
```diff
@@ -13,16 +13,18 @@
 use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 
-$article = $displayData['article'];
-$overlib = $displayData['overlib'];
-$legacy  = $displayData['legacy'];
+$article  = $displayData['article'];
```
Comment (Quy) [inline layouts/joomla/content/icons/edit.php]: Remove extra space.

**Comment by richard67** [comment]
@Quy Done.

**Comment by richard67** [comment]
@Quy When I check drone failure (which seems not to be caused by this PR) I see some red banner at the bottom of the drone page telling about some license to be exceeded.

**Comment by Quy** [comment]
Maybe @wilsonge or @Hackwar knows about the Drone's license issue.

**Comment by wilsonge** [comment]
Thanks!

**Comment by richard67** [comment]
Thanks too all testers and mergers 😄

**Pull request merged**